torch/_inductor/fx_passes/group_fusion.py

```diff
@@ -0,0 +1,135 @@
+import operator
+from typing import List
+
+import networkx
+import torch
+
+from ..pattern_matcher import (
+    CallFunctionVarArgs,
+    get_arg_value,
+    config_flag,
+    stable_topological_sort,
+)
+from torch._dynamo.utils import counters
+
+
+def get_nx_graph_from_fx_graph(graph: torch.fx.Graph) -> networkx.DiGraph:
+    G = networkx.DiGraph()
+    for node in graph.nodes:
+        G.add_node(node)
+        for user in node.users:
+            G.add_edge(node, user)
+    return G
+
+
+class GroupFusionPass:
+    def __init__(self, pattern, pair_check, replacement_fn, extra_check=lambda m: True):
+        self.pattern = pattern
+        self.extra_check = extra_check
+        self.pair_check = pair_check
+        self.replacement_fn = replacement_fn
+
+    def apply(self, graph):
+        target_nodes = []
+        for node in graph.nodes:
+            if m := self.pattern.match(node):
+                if self.extra_check(m):
+                    target_nodes.append(node)
+
+        seen_nodes = set()
+
+        for i, target_node in enumerate(target_nodes):
+            nodes_to_fuse = []
+            nx_graph = get_nx_graph_from_fx_graph(graph)
```

Comment:
computing this seems expensive, can it go outside the loop?

Reply:
So, actually no - because the graph can potentially change during each iteration of the loop. 

Initially, we tried finding all groupings and changing graph one-shot, but that didn't work as changing the graph affects future groupings.

One optimization I can do is not call this method if the graph isn't changing